### PR TITLE
backup: refine backup log for tracing backup tasks

### PR DESCRIPTION
**Pull request by YuJuncen**
### What problem does this PR solve? 
We send backup requests concurrently for better performance, but the log of them got mixed and make tracing the lifecycle of backing up a range harder. Like this:
image
(4 parallel backup tasks were started, and consequent `try backup` logs were mixed and hard to trace which task it was from.) 

### What is changed and how it works?
Some new utils `ContextWithFields` and `CL(ctx)` were introduced. The former one allows us to bind some fields into the contextual logger, and the latter one retrieves the contextual logger.
During backing up, each range split by `BuildBackupRangeAndSchema` would have a unique `range-sn`, this number would be added to the contextual logger and hence we can trace a backup task trivially with something like `grep`, as below:
image

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

### Release note

 - Refined the log for better tracing backup tasks.

### CODE REVIEW

**Comment by YuJuncen** [comment]
/build 
Why `github.com/pingcap/br/pkg/lightning/restore` failed... 🤔

File: pkg/backup/client.go [hunk 1]
```diff
@@ -840,6 +841,7 @@ func (bc *Client) handleFineGrained(
 // Stop receiving response if respFn returns error.
 func SendBackup(
 	ctx context.Context,
+	// the `storeID` seems only used for logging now, maybe we can remove it then?
```
Comment (3pointer) [inline pkg/backup/client.go:844]: Do you mean put `storeID` into context🤔

Comment (YuJuncen) [reply pkg/backup/client.go:844]: Yep, if it was only for logging...

**Comment by YuJuncen** [comment]
```
Error: RequestError: send request failed caused by: Put "http://127.0.0.1:24927/mybucket/br_s3/backupmeta": read tcp 127.0.0.1:60432->127.0.0.1:24927: read: connection reset by peer
```

When BR failed before the fail point triggered, `br_s3` may get stuck. (2hrs for this case)🤔

/run-integration-tests

**Review by 3pointer — APPROVED** [review]

File: pkg/backup/client.go [hunk 2]
```diff
@@ -423,10 +423,12 @@ func (bc *Client) BackupRanges(
 	// we collect all files in a single goroutine to avoid thread safety issues.
 	workerPool := utils.NewWorkerPool(concurrency, "Ranges")
 	eg, ectx := errgroup.WithContext(ctx)
-	for _, r := range ranges {
+	for id, r := range ranges {
+		id := id
```
Comment (joccau) [inline pkg/backup/client.go:427]: Is it redundant?

Comment (YuJuncen) [reply pkg/backup/client.go:427]: Nope, we need make `id` a local variable so the closure won't capture its change, which is a common trap in Go🤯.

See [this playground](https://play.golang.org/p/OyMjs3Bl4su).
```go
const max = 6

func main() {
	wg := new(sync.WaitGroup)
	wg.Add(max)
	for i := 0; i < max; i++ {
		go func() { fmt.Printf("%d", i); wg.Done() }()
	}
	wg.Wait()
	// output: 656666
}
```

**Review by joccau — APPROVED** [review]
LGTM

**Review by zwj-coder — APPROVED** [review]

**Comment by YuJuncen** [comment]
/merge

Even I'm wondering whether this need to be rewritten at TiDB.

**Pull request merged**